xapian-letor/ranker/svmranker.cc

```diff
@@ -0,0 +1,205 @@
+	    ++non_zero;
+    return non_zero;
+}
+
+void
+SVMRanker::train_model(const std::vector<Xapian::FeatureVector> & training_data)
+{
+    LOGCALL_VOID(API, "SVMRanker::train_model", training_data);
+    struct svm_parameter param;
+    param.svm_type = 4;         //nu-SVR
+    param.kernel_type = 0;      //linear Kernel
+    param.degree = 3;           //parameter for poly Kernel, default 3
+    param.gamma = 0;            //parameter for poly/rbf/sigmoid Kernel, default 1/num_features
+    param.coef0 = 0;            //parameter for poly/sigmoid Kernel, default 0
+    param.nu = 0.5;             //parameter for nu-SVC/one-class SVM/nu-SVR, default 0.5
+    param.cache_size = 100;     //default 40mb
+    param.C = 1;                //penalty parameter, default 1
+    param.eps = 1e-3;           //stopping criteria, default 1e-3
+    param.p = 0.1;              //parameter for e -SVR, default 0.1
+    param.shrinking = 1;        //use the shrinking heuristics
+    param.probability = 0;      //probability estimates
+    param.nr_weight = 0;        //parameter for C-SVCl
+    param.weight_label = NULL;  //parameter flor C-SVC
+    param.weight = NULL;        //parameter for c-SVC
+
+    int fvv_len = training_data.size();
+    int feature_cnt = -1;
+    if (fvv_len != 0) {
+	feature_cnt = training_data[0].get_fcount();
+    } else {
+	throw LetorInternalError("Training data is empty. Check training file.");
+    }
+
+    struct svm_problem prob;
+    // set the parameters for svm_problem
+    prob.l = fvv_len;
+    // feature vector
+    prob.x = new svm_node* [prob.l];
+    prob.y = new double [prob.l];
+
+    // Sparse storage; libsvm needs only non-zero features
+    for (int i = 0; i < fvv_len; ++i) {
+	prob.x[i] = new svm_node [get_non_zero_num(training_data[i]) + 1]; // one extra for default -1 at the end
+	int non_zero_flag = 0;
+	for (int k = 0; k < feature_cnt; ++k) {
+	    double fval = training_data[i].get_feature_value(k);
+	    if (fval != 0.0) {
+		prob.x[i][non_zero_flag].index = k;
+		prob.x[i][non_zero_flag].value = fval;
+		++non_zero_flag;
+	    }
+	}
+	prob.x[i][non_zero_flag].index = -1;
+	prob.x[i][non_zero_flag].value = -1;
+
+	prob.y[i] = training_data[i].get_label();
+    }
+
+    const char * error_msg;
+    error_msg = svm_check_parameter(&prob, &param);
+    if (error_msg)
+	throw LetorInternalError("svm_check_parameter failed: %s", error_msg);
+
+    struct svm_model * trainmodel;
+    trainmodel = svm_train(&prob, &param);
+
+    // Generate emporary file to extract model data
+    char templ[] = "/tmp/svmtemp.XXXXXX";
+    char * tempname = mktemp(templ);
+    svm_save_model(tempname, trainmodel);
+    // Read content of model to string
+    std::ifstream f(tempname);
+    std::ostringstream ss;
+    ss << f.rdbuf();
+    // Save model string
+    this->model_data = ss.str();
+    // Delete the temporary model file
+    std::remove(tempname);
+}
+
+void
+SVMRanker::save_model_to_metadata(const string & model_key)
```

Comment:
This is a protected method, so users aren't going to see it. Therefore it gets called by things that users called, which means (for instance) that it may be less likely that there's another writer to the database at the same time. If this is the case, then a comment here explaining why we're not bothering to handle exceptions would be appropriate. If not, I guess we should have something in the documentation for whatever public API call ends up calling this method saying that various exceptions might be raised, and in particular that you shouldn't have another writer on the same database while training.